File: java-examples/scenario1-web-app/service-optimization/before/StringHandlingService.java
package com.enterprise.optimization.service;

import org.springframework.stereotype.Service;
import java.util.List;
import java.util.ArrayList;

/**
 * Service for handling string operations in the application.
 * This implementation has several performance issues.
 */
@Service
public class StringHandlingService {

    /**
     * Concatenates a list of strings with a separator.
     * Inefficient implementation using String concatenation in a loop.
     */
    public String concatenateStrings(List<String> strings, String separator) {
        String result = "";
        for (String str : strings) {
            result = result + str + separator;
        }
        // Remove trailing separator
        if (result.length() > 0 && strings.size() > 0) {
            result = result.substring(0, result.length() - separator.length());
        }
        return result;
    }

    /**
     * Builds a complex string representation from an object.
     * Inefficient implementation using String concatenation.
     */
    public String buildComplexString(UserProfile profile) {
        String result = "";
        result = result + "User: " + profile.getUsername() + "\n";
        result = result + "Name: " + profile.getFirstName() + " " + profile.getLastName() + "\n";
        result = result + "Email: " + profile.getEmail() + "\n";
        result = result + "Joined: " + profile.getJoinDate() + "\n";
        result = result + "Status: " + profile.getStatus() + "\n";
        
        result = result + "Roles: ";
        for (String role : profile.getRoles()) {
            result = result + role + ", ";
        }
        
        // Remove trailing comma and space
        if (profile.getRoles().size() > 0) {
            result = result.substring(0, result.length() - 2);
        }
        
        return result;
    }

    /**
     * Searches for occurrences of a string within text.
     * Inefficient implementation using repeated String operations.
     */
    public List<Integer> findAllOccurrences(String text, String searchString) {
        List<Integer> positions = new ArrayList<>();
        String remainingText = text;
        int currentPosition = 0;
        
        while (remainingText.contains(searchString)) {
            int index = remainingText.indexOf(searchString);
            currentPosition += index;
            positions.add(currentPosition);
            
            // Move past this occurrence
            remainingText = remainingText.substring(index + searchString.length());
            currentPosition += searchString.length();
        }
        
        return positions;
    }

    /**
     * Replaces all occurrences of a string with another string.
     * Inefficient implementation using String operations in a loop.
     */
    public String replaceAllOccurrences(String text, String searchString, String replacement) {
        while (text.contains(searchString)) {
            int index = text.indexOf(searchString);
            text = text.substring(0, index) + replacement + text.substring(index + searchString.length());
        }
        return text;
    }

    /**
     * Dummy user profile class for the example
     */
    public static class UserProfile {
        private String username;
        private String firstName;
        private String lastName;
        private String email;
        private String joinDate;
        private String status;
        private List<String> roles;
        
        // Getters and setters
        public String getUsername() { return username; }
        public void setUsername(String username) { this.username = username; }
        
        public String getFirstName() { return firstName; }
        public void setFirstName(String firstName) { this.firstName = firstName; }
        
        public String getLastName() { return lastName; }
        public void setLastName(String lastName) { this.lastName = lastName; }
        
        public String getEmail() { return email; }
        public void setEmail(String email) { this.email = email; }
        
        public String getJoinDate() { return joinDate; }
        public void setJoinDate(String joinDate) { this.joinDate = joinDate; }
        
        public String getStatus() { return status; }
        public void setStatus(String status) { this.status = status; }
        
        public List<String> getRoles() { return roles; }
        public void setRoles(List<String> roles) { this.roles = roles; }
    }
} 

File: dotnet-examples/scenario1-web-app/service-optimization/before/StringHandlingService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Extensions.Logging;

namespace Enterprise.Web.Services
{
    /// <summary>
    /// Service for handling string operations in the application.
    /// This implementation has several performance issues.
    /// </summary>
    public class StringHandlingService
    {
        private readonly ILogger<StringHandlingService> _logger;

        public StringHandlingService(ILogger<StringHandlingService> logger)
        {
            _logger = logger;
        }

        /// <summary>
        /// Concatenates a list of strings with a separator.
        /// Inefficient implementation using string concatenation in a loop.
        /// </summary>
        public string ConcatenateStrings(IEnumerable<string> strings, string separator)
        {
            string result = "";
            foreach (var str in strings)
            {
                result = result + str + separator;
            }

            // Remove trailing separator
            if (result.Length > 0 && strings.Any())
            {
                result = result.Substring(0, result.Length - separator.Length);
            }

            return result;
        }

        /// <summary>
        /// Builds a complex string representation from an object.
        /// Inefficient implementation using string concatenation.
        /// </summary>
        public string BuildComplexString(UserProfile profile)
        {
            string result = "";
            result = result + "User: " + profile.Username + "\n";
            result = result + "Name: " + profile.FirstName + " " + profile.LastName + "\n";
            result = result + "Email: " + profile.Email + "\n";
            result = result + "Joined: " + profile.JoinDate + "\n";
            result = result + "Status: " + profile.Status + "\n";
            
            result = result + "Roles: ";
            foreach (var role in profile.Roles)
            {
                result = result + role + ", ";
            }
            
            // Remove trailing comma and space
            if (profile.Roles.Any())
            {
                result = result.Substring(0, result.Length - 2);
            }
            
            return result;
        }

        /// <summary>
        /// Searches for occurrences of a string within text.
        /// Inefficient implementation using repeated string operations.
        /// </summary>
        public List<int> FindAllOccurrences(string text, string searchString)
        {
            List<int> positions = new List<int>();
            string remainingText = text;
            int currentPosition = 0;
            
            while (remainingText.Contains(searchString))
            {
                int index = remainingText.IndexOf(searchString);
                currentPosition += index;
                positions.Add(currentPosition);
                
                // Move past this occurrence
                remainingText = remainingText.Substring(index + searchString.Length);
                currentPosition += searchString.Length;
            }
            
            return positions;
        }

        /// <summary>
        /// Replaces all occurrences of a string with another string.
        /// Inefficient implementation using string operations in a loop.
        /// </summary>
        public string ReplaceAllOccurrences(string text, string searchString, string replacement)
        {
            while (text.Contains(searchString))
            {
                int index = text.IndexOf(searchString);
                text = text.Substring(0, index) + replacement + text.Substring(index + searchString.Length);
            }
            return text;
        }
    }

    /// <summary>
    /// Dummy user profile class for the example
    /// </summary>
    public class UserProfile
    {
        public string Username { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string Email { get; set; }
        public DateTime JoinDate { get; set; }
        public string Status { get; set; }
        public List<string> Roles { get; set; } = new List<string>();
    }
} 